### PR TITLE
apply cluster-aware overrides from Globalization

### LINKED ISSUE #333: clusterAffinity of Globalization not working
#### What happened:

I want to override different configs for a subscription  according to different labels of mcls, so I  created two Globalization and use `matchLabels` of `clusterAffinity` field to query different mcls , but `clusterAffinity` does  not  work, all overrides of Globalizations merged in target `description`

#### Environment:
- Clusternet version:
    - Clusternet-agent version (user `clusternet-agent --version=json`):0.9.1
    - Clusternet-hub version (user `clusternet-hub --version=json`): 0.9.1
- Kubernetes version (use `kubectl version`): k3s  1.22
- Cloud provider or hardware configuration: 
- OS (e.g: `cat /etc/os-release`): ubuntu 20.04

### PR DESCRIPTION
**Pull request by jasine**
Signed-off-by: jasine <jasinechen@gmail.com>

#### What type of PR is this?
kind/feature

#### What this PR does / why we need it:

apply cluster-aware overrides from `Globalization`

#### Which issue(s) this PR fixes:

Fixes #333

### CODE REVIEW

File: pkg/scheduler/algorithm/generic.go [hunk 1]
```diff
@@ -58,7 +57,7 @@ const (
 var ErrNoClustersAvailable = fmt.Errorf("no clusters available to schedule subscriptions")
 
 type genericScheduler struct {
-	cache                       schedulercache.Cache
+	cache                       utils.MclsCache
```
Comment (dixudx) [inline pkg/scheduler/algorithm/generic.go:60]: Don't modify this.

The cache here is for `scheduler`.

**Review by dixudx — CHANGES_REQUESTED** [review]
`pkg/hub/localizer/localizer.go` is the only place for making changes.

File: pkg/scheduler/algorithm/generic.go (hunk 1 above)
Comment (jasine) [reply pkg/scheduler/algorithm/generic.go:60]: then the same file `cache` have to write twice, I moved it to util to share between hub and scheduler

File: pkg/hub/localizer/localizer.go [hunk 2]
```diff
@@ -331,12 +334,26 @@ func (l *Localizer) getOverrides(namespace string, feed appsapi.Feed) ([]appsapi
...
+	feedGlobs, err := l.globLister.List(labels.SelectorFromSet(labels.Set{
 		string(uid): feed.Kind,
 	}))
 	if err != nil {
 		return nil, err
 	}
+
+	globs := make([]*appsapi.Globalization, 0)
+	for _, glob := range feedGlobs {
+		clusters, err := l.mclsCache.ListByNamespace(namespace, glob.Spec.ClusterAffinity)
```
Comment (dixudx) [inline pkg/hub/localizer/localizer.go:346]: `Description` is a namespace-scoped resource. Here we need to verify whether the selector in `Globalization` matches the `ManagedCluster` in the same namespace with `Description`.

File: pkg/hub/localizer/localizer.go [hunk 3]
```diff
@@ -331,12 +334,26 @@ func (l *Localizer) getOverrides(namespace string, feed appsapi.Feed) ([]appsapi
...
+		if err != nil {
+			return nil, err
+		}
+		if len(clusters) == 0 {
```
Comment (dixudx) [inline pkg/hub/localizer/localizer.go:350]: This is not correct. What if the `clusterAffinity` does match a group of clusters, but none of them are in the namespace that we're processing.

File: pkg/hub/localizer/localizer.go [hunk 4]
```diff
@@ -67,6 +67,8 @@ type Localizer struct {
 	chartCallback    func(*appsapi.HelmChart) error
 	manifestCallback func(*appsapi.Manifest) error
 
+	mclsCache utils.MclsCache
```
Comment (dixudx) [inline pkg/hub/localizer/localizer.go:70]: Why not using a lister for `ManagedCluster` directly ?

**Review by dixudx — APPROVED** [review]
Good work! 👍🏻
Thanks for your contribution.

**Pull request merged**